File: packages/expo-contacts/android/src/main/java/expo/modules/contacts/models/ExtraNameModel.java
package expo.modules.contacts.models;

import android.database.Cursor;
import android.provider.ContactsContract;

import java.util.Map;

import expo.modules.contacts.EXColumns;

import static android.provider.ContactsContract.*;

public class ExtraNameModel extends BaseModel {

    @Override
    public String getContentType() {
        return CommonDataKinds.Nickname.CONTENT_ITEM_TYPE;
    }

    @Override
    public String getDataAlias() {
        return "value";
    }

    @Override
    public int mapStringToType(String label) {
        switch (label) {
        case "default":
            return CommonDataKinds.Nickname.TYPE_DEFAULT;
        case "initials":
            return CommonDataKinds.Nickname.TYPE_INITIALS;
        case "maidenName":
            return CommonDataKinds.Nickname.TYPE_MAIDEN_NAME;
        case "shortName":
            return CommonDataKinds.Nickname.TYPE_SHORT_NAME;
        case "otherName":
            return CommonDataKinds.Nickname.TYPE_OTHER_NAME;
        default:
            return CommonDataKinds.Nickname.TYPE_CUSTOM;
        }
    }

    @Override
    protected String getLabelFromCursor(Cursor cursor) {
        String label = super.getLabelFromCursor(cursor);
        if (label != null)
            return label;
        switch (cursor.getInt(cursor.getColumnIndex(EXColumns.TYPE))) {
        case CommonDataKinds.Nickname.TYPE_DEFAULT:
            return "nickname";
        case CommonDataKinds.Nickname.TYPE_INITIALS:
            return "initials";
        case CommonDataKinds.Nickname.TYPE_MAIDEN_NAME:
            return "maidenName";
        case CommonDataKinds.Nickname.TYPE_SHORT_NAME:
            return "shortName";
        case CommonDataKinds.Nickname.TYPE_OTHER_NAME:
            return "otherName";
        default:
            return "unknown";
        }
    }

    @Override
    public void fromMap(Map<String, Object> readableMap) {
        super.fromMap(readableMap);
        // TODO: Evan: Decode contact data
    }
}

File: packages/expo-contacts/android/src/main/java/expo/modules/contacts/models/ExtraNameModel.kt
package expo.modules.contacts.models

import android.database.Cursor
import android.provider.ContactsContract.CommonDataKinds
import expo.modules.contacts.Columns

class ExtraNameModel : BaseModel() {
  override val contentType: String = CommonDataKinds.Nickname.CONTENT_ITEM_TYPE
  override val dataAlias: String = "value"

  override fun mapStringToType(label: String?): Int {
    return when (label) {
      "default" -> CommonDataKinds.Nickname.TYPE_DEFAULT
      "initials" -> CommonDataKinds.Nickname.TYPE_INITIALS
      "maidenName" -> CommonDataKinds.Nickname.TYPE_MAIDEN_NAME
      "shortName" -> CommonDataKinds.Nickname.TYPE_SHORT_NAME
      "otherName" -> CommonDataKinds.Nickname.TYPE_OTHER_NAME
      else -> CommonDataKinds.Nickname.TYPE_CUSTOM
    }
  }

  override fun getLabelFromCursor(cursor: Cursor): String {
    val label = super.getLabelFromCursor(cursor)
    return label
      ?: when (cursor.getInt(cursor.getColumnIndexOrThrow(Columns.TYPE))) {
        CommonDataKinds.Nickname.TYPE_DEFAULT -> "nickname"
        CommonDataKinds.Nickname.TYPE_INITIALS -> "initials"
        CommonDataKinds.Nickname.TYPE_MAIDEN_NAME -> "maidenName"
        CommonDataKinds.Nickname.TYPE_SHORT_NAME -> "shortName"
        CommonDataKinds.Nickname.TYPE_OTHER_NAME -> "otherName"
        else -> "unknown"
      }
  }
}
